### PR TITLE
#5: Add JBrowse tab & launch button

### PR DESCRIPTION
**Pull request by samrichca**
- related dms-jbrowse PR: https://github.com/overture-stack/dms-jbrowse-components/pull/16
- add tabs & tabs context
- move tooltip to its own file
- add JBrowse tab
- add JBrowse launch button

### CODE REVIEW

File: components/pages/user/ApiTokenInfo.tsx [hunk 1]
```diff
@@ -49,41 +47,6 @@ interface ApiToken {
   scope: string[];
 }
 
-interface CustomTooltipProps extends TooltipProps {
```
Comment (samrichca) [inline components/pages/user/ApiTokenInfo.tsx:52]: moved tooltip to its own file

File: package.json [hunk 2]
```diff
@@ -50,7 +50,8 @@
 		"trailingComma": "all",
 		"semi": true,
 		"singleQuote": true,
-		"useTabs": true
+		"useTabs": false,
```
Comment (samrichca) [inline package.json:53]: not a personal preference - without these settings, I was seeing unintentional formatting changes (spaces to tabs, 4 wide)

File: components/Tooltip.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
+/*
+ *
```
Comment (samrichca) [inline components/Tooltip.tsx:2]: was previously in another file, see below

File: components/pages/explorer/RepositoryContent.tsx [hunk 4]
```diff
@@ -0,0 +1,265 @@
+/*
+ *
```
Comment (samrichca) [inline components/pages/explorer/RepositoryContent.tsx:2]: renamed RepoTable to RepositoryContent

File: components/pages/explorer/ActionBar.tsx [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  margin-bottom: 0.3rem;
+`;
+
+const ActionBar = () => {
+  const theme = useTheme();
+  const { selectedRows } = useTableContext({
+    callerName: 'Repository - ActionBar',
+  });
+  const { activeTab, handleChangeTab, handleOpenTab, openTabs } = useTabsContext();
+  const enableJbrowse = !!selectedRows.length;
```
Comment (samrichca) [inline components/pages/explorer/ActionBar.tsx:46]: this is temporary. if any files are selected you can open the JBrowse tab.

File: components/pages/explorer/RepositoryContent.tsx [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    },
+  },
+});
+
+const ContentDisplay = () => {
+  const { activeTab } = useTabsContext();
+  const { isLoading } = useTableContext();
+  const theme = useTheme();
+
+  return activeTab === RepositoryTabNames.FILES ? (
```
Comment (justincorrigible) [inline components/pages/explorer/RepositoryContent.tsx:169]: Seems like this logic could potentially grow at some point, so preemptively I'd suggest a `switch` rather than a ternary chain?

File: components/pages/explorer/TablePagination.tsx [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+      >
+        <MaxRowsSelector
+          css={css`
+            ${textCss}
+            margin-left: 0.3rem;
+            .Spinner {
+              justify-content: space-between;
+              width: 65%;
+            }
+          `}
```
Comment (justincorrigible) [inline components/pages/explorer/TablePagination.tsx:63]: Fun fact
```suggestion
          css={[
            css`
              margin-left: 0.3rem;

              .Spinner {
                justify-content: space-between;
                width: 65%;
              }
            `,
            textCss
        ]}
```

File: components/pages/explorer/TablePagination.tsx [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        />
+        <CountDisplay
+          css={css`
+            ${textCss}
```
Comment (justincorrigible) [inline components/pages/explorer/TablePagination.tsx:67]: Same pattern as mentioned above would be better here.

File: components/pages/explorer/TabsContext.tsx [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  handleOpenTab: (tab: TabRecord) => void;
+  handleChangeTab: (tabName: string) => void;
+  handleCloseTab: (tabName: string) => void;
+  openTabs: TabRecord[];
+}
+
+export const TabsContext = createContext<TabsContextInterface>({
+  activeTab: null,
+  openTabs: [],
+} as unknown as TabsContextInterface);
```
Comment (justincorrigible) [inline components/pages/explorer/TabsContext.tsx:41]: is the `as unknown` step necessary here?

File: package.json (hunk 2 above)
Comment (justincorrigible) [reply package.json:53]: This change was made on purpose, as the new industry standard is to use tabs to facilitate accessibility for our visually-disabled colleagues. While I realise that complicates reviews, it is a good change worth the reviewer's effort.

```suggestion
		"useTabs": true,
```

File: tsconfig.json [hunk 10]
```diff
@@ -1,39 +1,39 @@
 {
```
Comment (justincorrigible) [inline tsconfig.json:1]: Please revert this change, for the reason explained before (tabs are accessible).

**Review by justincorrigible — CHANGES_REQUESTED** [review]
Looks pretty good overall! 💪 
Left a suggestion on how to do CSS mixins, and a request to not change the indentation rules (tabs vs spaces).

File: components/pages/explorer/TablePagination.tsx (hunk 7 above)
Comment (samrichca) [reply components/pages/explorer/TablePagination.tsx:63]: good to know!

File: components/pages/explorer/TabsContext.tsx (hunk 9 above)
Comment (justincorrigible) [reply components/pages/explorer/TabsContext.tsx:41]: 🤔 fairly sure I've done that in Arranger, and this middle-step is not required (https://github.com/overture-stack/arranger/blob/develop/modules/components/src/DataContext/index.tsx#L17)
Wonder what I'm doing differently there

File: components/pages/explorer/RepositoryContent.tsx (hunk 4 above)
Comment (justincorrigible) [reply components/pages/explorer/RepositoryContent.tsx:2]: Weren't you arguing about scope earlier? 😆

File: components/pages/explorer/TabsContext.tsx (hunk 9 above)
Comment (samrichca) [reply components/pages/explorer/TabsContext.tsx:41]: Conversion of type (activeTab, openTabs)
 to type (TabsContextInterface)
 may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to unknown first.

TableContext has the unknown step as well but I'm not sure why

File: components/pages/explorer/RepositoryContent.tsx (hunk 4 above)
Comment (samrichca) [reply components/pages/explorer/RepositoryContent.tsx:2]: it's not a table anymore 💯

Comment (justincorrigible) [reply components/pages/explorer/RepositoryContent.tsx:2]: I see!!! makes sense now

File: components/pages/explorer/TabsContext.tsx (hunk 9 above)
Comment (justincorrigible) [reply components/pages/explorer/TabsContext.tsx:41]: Looks like it doesn't like `never[]` vs `TabRecord[]`

the solution is to make `openTabs: TabRecord[] | [];` in line 35

Comment (samrichca) [reply components/pages/explorer/TabsContext.tsx]: nice, updated

**Review by justincorrigible — APPROVED** [review]

**Comment by justincorrigible** [comment]
Thanks for all the hard work, @samrichca!
